bytecomp/matching.ml

```diff
@@ -13,7 +13,56 @@
 (*                                                                        *)
 (**************************************************************************)
 
-(* Compilation of pattern matching *)
+(* Compilation of pattern matching
+
+   Based upon Lefessant-Maranget ``Optimizing Pattern-Matching'' ICFP'2001.
+
+   A previous version was based on Peyton-Jones, ``The Implementation of
+   functional programming languages'', chapter 5.
+
+   # Overview of the implementation
+
+     (split_and_precompile)
+   We first split the matching along its first column -- simplifying pattern
+   heads in the process --, so that we obtain a sorted list of pattern matchings
+   (or "pms").
+   An invariant that is enforced in this step is that for each of the resulting
+   pms, its rows match disjoint sets of values.
+   The idea being that if a pm in the list fails to match the scrutiny, one can
+   jump to one of the nexts pms in the list and try again.
```

Comment:
I have read those two sentences/paragraphs several times and I don't find them very clear:

1. I'm not convinced that it is actually true that two rows of a single pm match disjoint sets of values. This is true of the or-matrix, but what about the `yes` matrix in `split_or`?
2. I'm not sure what you mean with "if a pm in the list fails to match the scrutiny"; naively one would expect this to mean "none of the rows of the pm match the scrutiny", but then this sentence is obvious (and not a result of the invariant you mention before, but something that always holds): the semantics of the ordered list of pms is that you try them in order, right?

Reply:
I suspect that a description of this part should probably say something about or-patterns. (Or is it possible to abstract away the details at this step and only talk about it in `split_or` and `precompile_or`? Where are we having this comment with pictures of the compilation technique for or-patterns?)

Reply:
> but what about the `yes` matrix in `split_or`?

Here you're talking about intermediate results/steps of what this collection of functions does; my intention was to speak only of the "final results".
Which is not to say that the invariant I stated necessarily holds (I wrote this paragraph purely from memory), in fact I believe you're right and that it doesn't.
The invariant that holds is the one that we had written in the notes:
> ```
>      Luc: on peut dire que l'hypothèse faite par divide, c'est que sur la
>      matrice reçue en argument, on peut décider de l'égalité entre deux
>      têtes de pattern arbitraires de la première colonne.
>        (soit elles sont syntaxiquement égales, soit on les sait disjointes)
>      
>      Thomas: Donc il faut que split(_and_precompile) enforce cette invariant
>      dans les matrices qu'il produit.
> ```


> but then this sentence is obvious (and not a result of the invariant you mention before, but something that always holds): the semantics of the ordered list of pms is that you try them in order, right?

You're right. Clearly I did not express my thoughts correctly. But what I wanted to say was wrong (based on a incorrect memory of the invariant). I'll rewrite that part.

> I suspect that a description of this part should probably say something about or-patterns.

I feel differently. I think this is a detail that does not need to be exposed at such a high-level.
We have comments around the specific functions, which I think is the right place, and is enough.

> Where are we having this comment with pictures of the compilation technique for or-patterns?

I imagined it attached to `split_or` and `precompile_or`

Reply:
Your idea to leave or-handling as an internal detail of the split&precompile passes is a good idea, if we can write down a good specification of the result without talking about it.